File: src/remote_browser_page.py
# remote_browser_page.py
#
# Copyright 2026 Unknown
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
# SPDX-License-Identifier: GPL-3.0-or-later

from gi.repository import Adw, Gtk, GLib
from . import keyring
from .webdav import WebDAVClient


@Gtk.Template(resource_path='/io/github/nico359/cloudsend/remote_browser_page.ui')
class CloudsendRemoteBrowserPage(Adw.NavigationPage):
    __gtype_name__ = 'CloudsendRemoteBrowserPage'

    browser_stack = Gtk.Template.Child()
    folders_listbox = Gtk.Template.Child()
    select_button = Gtk.Template.Child()
    new_folder_button = Gtk.Template.Child()
    error_status = Gtk.Template.Child()

    def __init__(self, current_path='/', on_selected=None, **kwargs):
        """
        current_path: initial remote path to browse
        on_selected: callback(selected_remote_path) called when user taps Select
        """
        super().__init__(**kwargs)
        self._current_path = current_path.rstrip('/') or '/'
        self._on_selected = on_selected
        self._client = None

        self.select_button.connect('clicked', self._on_select_clicked)
        self.new_folder_button.connect('clicked', self._on_new_folder_clicked)

        self._setup_client()
        self._load_directory()

    def _setup_client(self):
        creds = keyring.load_credentials()
        if creds:
            self._client = WebDAVClient(
                creds['server_url'], creds['username'], creds['app_password']
            )

    def _load_directory(self):
        """Load the current remote path contents."""
        self.set_title(self._current_path or '/')
        self.browser_stack.set_visible_child_name('loading')

        while True:
            row = self.folders_listbox.get_row_at_index(0)
            if row is None:
                break
            self.folders_listbox.remove(row)

        if not self._client:
            self.error_status.set_description("No account configured")
            self.browser_stack.set_visible_child_name('error')
            return

        self._client.list_directory(self._current_path, self._on_list_result)

    def _on_list_result(self, success, result):
        def update():
            if not success:
                self.error_status.set_description(str(result))
                self.browser_stack.set_visible_child_name('error')
                return

            if self._current_path and self._current_path != '/':
                parent_row = Adw.ActionRow()
                parent_row.set_title("📁 ..")
                parent_row.set_subtitle("Go up")
                parent_row.set_activatable(True)
                parent_row.connect('activated', self._on_go_up)
                self.folders_listbox.append(parent_row)

            # Only show directories (this is a folder picker)
            dirs = sorted(
                (item for item in result if item['is_dir']),
                key=lambda d: d['name'].lower(),
            )

            for item in dirs:
                row = Adw.ActionRow()
                row.set_title(f"📁 {item['name']}")
                row.set_activatable(True)
                row.connect('activated', self._on_folder_activated, item['name'])
                self.folders_listbox.append(row)

            self.browser_stack.set_visible_child_name('content')

        GLib.idle_add(update)

    def _on_folder_activated(self, row, folder_name):
        """Navigate into a subfolder."""
        if self._current_path == '/':
            self._current_path = '/' + folder_name
        else:
            self._current_path = self._current_path.rstrip('/') + '/' + folder_name
        self._load_directory()

    def _on_go_up(self, row):
        """Navigate to parent directory."""
        if '/' in self._current_path.rstrip('/'):
            self._current_path = self._current_path.rstrip('/').rsplit('/', 1)[0]
        if not self._current_path:
            self._current_path = '/'
        self._load_directory()

    def _on_select_clicked(self, button):
        """Confirm selection and pop the page."""
        if self._on_selected:
            self._on_selected(self._current_path)
        nav = self.get_ancestor(Adw.NavigationView)
        if nav:
            nav.pop()

    def _on_new_folder_clicked(self, button):
        """Show dialog to create a new folder in the current directory."""
        dialog = Adw.AlertDialog(
            heading="New Folder",
            body="Enter a name for the new folder:",
        )

        entry = Gtk.Entry()
        entry.set_placeholder_text("Folder name")
        dialog.set_extra_child(entry)

        dialog.add_response("cancel", "Cancel")
        dialog.add_response("create", "Create")
        dialog.set_response_appearance("create", Adw.ResponseAppearance.SUGGESTED)
        dialog.set_default_response("create")

        def on_response(dlg, response):
            if response == "create":
                name = entry.get_text().strip()
                if name:
                    path = self._current_path.rstrip('/') + '/' + name
                    self._client.create_directory(
                        path,
                        lambda ok, err: GLib.idle_add(self._on_folder_created, ok, err),
                    )

        dialog.connect('response', on_response)
        dialog.present(self.get_ancestor(Gtk.Window))

    def _on_folder_created(self, success, error):
        """Refresh the listing after creating a folder."""
        if success:
            self._load_directory()
            toast = Adw.Toast(title="Folder created")
            win = self.get_ancestor(Adw.ApplicationWindow)
            if hasattr(win, '_toast_overlay'):
                win._toast_overlay.add_toast(toast)
        else:
            dialog = Adw.AlertDialog(
                heading="Folder Creation Failed",
                body=str(error),
            )
            dialog.add_response("close", "Close")
            dialog.set_default_response("close")
            dialog.present(self.get_ancestor(Gtk.Window))

File: src/remote_browser_page.rs
use adw::prelude::*;
use adw::subclass::prelude::*;
use gtk::glib;
use std::cell::RefCell;

use crate::keyring::Credentials;
use crate::webdav::{RemoteItem, WebDAVClient};
use crate::window::SimplesyncWindow;

mod imp {
    use super::*;

    #[derive(Default, gtk::CompositeTemplate)]
    #[template(resource = "/io/github/nico359/simplesync/remote_browser_page.ui")]
    pub struct SimplesyncRemoteBrowserPage {
        #[template_child]
        pub select_button: TemplateChild<gtk::Button>,
        #[template_child]
        pub content_stack: TemplateChild<gtk::Stack>,
        #[template_child]
        pub path_label: TemplateChild<gtk::Label>,
        #[template_child]
        pub folder_list: TemplateChild<gtk::ListBox>,
        #[template_child]
        pub new_folder_button: TemplateChild<gtk::Button>,
        #[template_child]
        pub error_status: TemplateChild<adw::StatusPage>,

        pub window: RefCell<Option<SimplesyncWindow>>,
        pub current_path: RefCell<String>,
        pub creds: RefCell<Option<Credentials>>,
        pub on_select: RefCell<Option<Box<dyn Fn(String) + 'static>>>,
    }

    impl std::fmt::Debug for SimplesyncRemoteBrowserPage {
        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
            f.debug_struct("SimplesyncRemoteBrowserPage").finish()
        }
    }

    #[glib::object_subclass]
    impl ObjectSubclass for SimplesyncRemoteBrowserPage {
        const NAME: &'static str = "SimplesyncRemoteBrowserPage";
        type Type = super::SimplesyncRemoteBrowserPage;
        type ParentType = adw::NavigationPage;

        fn class_init(klass: &mut Self::Class) {
            klass.bind_template();
        }

        fn instance_init(obj: &glib::subclass::InitializingObject<Self>) {
            obj.init_template();
        }
    }

    impl ObjectImpl for SimplesyncRemoteBrowserPage {}
    impl WidgetImpl for SimplesyncRemoteBrowserPage {}
    impl NavigationPageImpl for SimplesyncRemoteBrowserPage {}
}

glib::wrapper! {
    pub struct SimplesyncRemoteBrowserPage(ObjectSubclass<imp::SimplesyncRemoteBrowserPage>)
        @extends gtk::Widget, adw::NavigationPage;
}

impl SimplesyncRemoteBrowserPage {
    pub fn new() -> Self {
        glib::Object::builder().build()
    }

    pub fn set_window<F: Fn(String) + 'static>(
        &self,
        window: &SimplesyncWindow,
        creds: &Credentials,
        on_select: F,
    ) {
        self.imp().window.replace(Some(window.clone()));
        self.imp().on_select.replace(Some(Box::new(on_select)));
        self.imp().current_path.replace("/".to_string());
        self.imp().creds.replace(Some(creds.clone()));

        self.setup_signals();
        self.load_directory("/");
    }

    fn window(&self) -> SimplesyncWindow {
        self.imp().window.borrow().clone().expect("Window not set")
    }

    fn make_client(&self) -> Option<WebDAVClient> {
        let creds = self.imp().creds.borrow();
        creds.as_ref().map(|c| WebDAVClient::new(&c.server_url, &c.username, &c.app_password))
    }

    fn setup_signals(&self) {
        let page = self.clone();
        self.imp().select_button.connect_clicked(move |_| {
            let path = page.imp().current_path.borrow().clone();
            if let Some(ref callback) = *page.imp().on_select.borrow() {
                callback(path);
            }
            page.window().navigation_view().pop();
        });

        let page = self.clone();
        self.imp().new_folder_button.connect_clicked(move |_| {
            page.show_new_folder_dialog();
        });
    }

    fn load_directory(&self, path: &str) {
        self.imp().content_stack.set_visible_child_name("loading");
        self.imp().current_path.replace(path.to_string());
        self.imp().path_label.set_text(path);

        let client = match self.make_client() {
            Some(c) => c,
            None => return,
        };
        let path_owned = path.to_string();

        let (tx, rx) = std::sync::mpsc::channel::<Result<Vec<RemoteItem>, String>>();
        std::thread::spawn(move || {
            let result = client.list_directory(&path_owned).map_err(|e| e.to_string());
            let _ = tx.send(result);
        });

        let page = self.clone();
        let path_for_cb = path.to_string();
        glib::timeout_add_local(std::time::Duration::from_millis(100), move || {
            match rx.try_recv() {
                Ok(Ok(items)) => {
                    page.populate_list(&path_for_cb, items);
                    page.imp().content_stack.set_visible_child_name("content");
                    glib::ControlFlow::Break
                }
                Ok(Err(e)) => {
                    page.imp().error_status.set_description(Some(&e));
                    page.imp().content_stack.set_visible_child_name("error");
                    glib::ControlFlow::Break
                }
                Err(std::sync::mpsc::TryRecvError::Empty) => glib::ControlFlow::Continue,
                Err(_) => glib::ControlFlow::Break,
            }
        });
    }

    fn populate_list(&self, current_path: &str, items: Vec<RemoteItem>) {
        let list = &self.imp().folder_list;

        while let Some(child) = list.first_child() {
            list.remove(&child);
        }

        // Parent directory row
        if current_path != "/" && !current_path.is_empty() {
            let parent = std::path::Path::new(current_path)
                .parent()
                .map(|p| {
                    let s = p.to_string_lossy().to_string();
                    if s.is_empty() { "/".to_string() } else { s }
                })
                .unwrap_or_else(|| "/".to_string());

            let row = adw::ActionRow::builder()
                .title("..")
                .subtitle("Parent directory")
                .activatable(true)
                .build();
            row.add_prefix(&gtk::Image::from_icon_name("go-up-symbolic"));

            let page = self.clone();
            row.connect_activated(move |_| {
                page.load_directory(&parent);
            });
            list.append(&row);
        }

        let mut dirs: Vec<_> = items.into_iter().filter(|i| i.is_dir).collect();
        dirs.sort_by(|a, b| a.name.to_lowercase().cmp(&b.name.to_lowercase()));

        for item in dirs {
            let child_path = if current_path == "/" || current_path.is_empty() {
                format!("/{}", item.name)
            } else {
                format!("{}/{}", current_path.trim_end_matches('/'), item.name)
            };

            let row = adw::ActionRow::builder()
                .title(&item.name)
                .activatable(true)
                .build();
            row.add_prefix(&gtk::Image::from_icon_name("folder-symbolic"));

            let page = self.clone();
            row.connect_activated(move |_| {
                page.load_directory(&child_path);
            });
            list.append(&row);
        }
    }

    fn show_new_folder_dialog(&self) {
        let dialog = adw::AlertDialog::builder()
            .heading("New Folder")
            .body("Enter a name for the new folder")
            .build();
        dialog.add_responses(&[("cancel", "Cancel"), ("create", "Create")]);
        dialog.set_response_appearance("create", adw::ResponseAppearance::Suggested);
        dialog.set_default_response(Some("create"));

        let entry = gtk::Entry::builder()
            .placeholder_text("Folder name")
            .build();
        dialog.set_extra_child(Some(&entry));

        let page = self.clone();
        dialog.connect_response(None, move |_, response| {
            if response == "create" {
                let name = entry.text().trim().to_string();
                if name.is_empty() {
                    return;
                }

                let current = page.imp().current_path.borrow().clone();
                let new_path = if current == "/" || current.is_empty() {
                    format!("/{}", name)
                } else {
                    format!("{}/{}", current.trim_end_matches('/'), name)
                };

                let client = match page.make_client() {
                    Some(c) => c,
                    None => return,
                };

                let (tx, rx) = std::sync::mpsc::channel::<Result<(), String>>();
                std::thread::spawn(move || {
                    let result = client.create_directory(&new_path).map_err(|e| e.to_string());
                    let _ = tx.send(result);
                });

                let page_clone = page.clone();
                let current_clone = current.clone();
                glib::timeout_add_local(std::time::Duration::from_millis(100), move || {
                    match rx.try_recv() {
                        Ok(Ok(())) => {
                            page_clone.load_directory(&current_clone);
                            glib::ControlFlow::Break
                        }
                        Ok(Err(e)) => {
                            page_clone.window().show_toast(&format!("Error: {}", e));
                            glib::ControlFlow::Break
                        }
                        Err(std::sync::mpsc::TryRecvError::Empty) => glib::ControlFlow::Continue,
                        Err(_) => glib::ControlFlow::Break,
                    }
                });
            }
        });

        dialog.present(Some(&self.window()));
    }
}
